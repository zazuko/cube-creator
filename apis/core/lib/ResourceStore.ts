import type StreamClient from 'sparql-http-client/StreamClient'
import { NamedNode, Quad, Term } from 'rdf-js'
import cf, { GraphPointer } from 'clownface'
import $rdf from 'rdf-ext'
import DatasetExt from 'rdf-ext/lib/Dataset'
import { CONSTRUCT, INSERT } from '@tpluscode/sparql-builder'
import TermMap from '@rdfjs/term-map'
import { hydra, rdf } from '@tpluscode/rdf-ns-builders'
import { warn } from '@hydrofoil/labyrinth/lib/logger'
import { sparql } from '@tpluscode/rdf-string'
import TermSet from '@rdfjs/term-set'
import RdfResource, { RdfResourceCore } from '@tpluscode/rdfine/RdfResource'
import { NotFoundError } from './errors'

interface ResourceCreationOptions {
  implicitlyDereferencable?: boolean
}

interface GetOptions {
  allowMissing?: boolean
}

/**
 * Represents an in-memory resource store where each resource is a graph pointer.
 * When the `save` method is called those resources are persisted in their respective
 * named graphs in a triple store
 */
export interface ResourceStore {
  /**
   * Gets all triples from a named graph and returns a graph pointer to a resource
   * identified by the same URI
   */
  get(id: string | Term | undefined, opts?: GetOptions): Promise<GraphPointer<NamedNode, DatasetExt>>
  get(id: string | Term | undefined, opts: { allowMissing: true }): Promise<GraphPointer<NamedNode, DatasetExt> | undefined>

  getResource<T extends RdfResourceCore>(id: string | Term | undefined, opts?: GetOptions): Promise<T>
  getResource<T extends RdfResourceCore>(id: string | Term | undefined, opts: { allowMissing: true }): Promise<T> | undefined

  /**
   * Creates a new resource a puts in the in-memory store
   */
  create(id: Term, options?: ResourceCreationOptions): GraphPointer<NamedNode, DatasetExt>

  /**
   * Create a new collection member, initialized according to `hydra:manages` entries
   */
  createMember(collection: Term, id: NamedNode, options?: ResourceCreationOptions): Promise<GraphPointer<NamedNode, DatasetExt>>

  /**
   * Replaces the resources in the triple store by inserting default graph
   * quads from each graph pointer into a named graph identified by that
   * graph pointer
   */
  save(): Promise<void>

  /**
   * Removes the named graph from the triple store
   */
  delete(id: NamedNode): void
}

interface TripleStoreFacade {
  loadResource(term: NamedNode): Promise<GraphPointer<NamedNode, DatasetExt> | undefined>
  writeChanges(resources: Map<Term, GraphPointer>, deletedResources: Iterable<NamedNode>): Promise<void>
}

function toTriple({ subject, predicate, object }: Quad) {
  return $rdf.quad(subject, predicate, object)
}

class SparqlStoreFacade implements TripleStoreFacade {
  private readonly __client: StreamClient

  constructor(client: StreamClient) {
    this.__client = client
  }

  async loadResource(term: NamedNode): Promise<GraphPointer<NamedNode, DatasetExt>> {
    const stream = await CONSTRUCT`?s ?p ?o`
      .WHERE`GRAPH ${term} { ?s ?p ?o }`
      .execute(this.__client.query)

    return cf({ dataset: await $rdf.dataset().import(stream), term })
  }

  deleteQuery(terms: Iterable<NamedNode>) {
    let deleteQuery = sparql``
    for (const id of terms) {
      deleteQuery = sparql`${deleteQuery}DROP SILENT GRAPH ${id};\n`
    }

    return deleteQuery
  }

  writeChanges(resources: Map<NamedNode, GraphPointer>, deletedResources: Iterable<NamedNode>): Promise<void> {
    const deleteGraphs = this.deleteQuery([...deletedResources, ...resources.keys()])
    const insertData = [...resources.entries()].reduce((insert, [graph, { dataset }]) => {
      return insert.DATA`GRAPH ${graph} {
        ${[...dataset].map(toTriple)}
      }`
    }, INSERT.DATA``)._getTemplateResult()

    const query = sparql`${deleteGraphs}\n${insertData}`
    return this.__client.query.update(query.toString())
  }
}

export default class implements ResourceStore {
  private readonly __storage: TripleStoreFacade
  private readonly __resources: TermMap<NamedNode, GraphPointer<NamedNode, DatasetExt>>
  private readonly __deletedGraphs: TermSet<NamedNode>

  constructor(clientOrStore: StreamClient | TripleStoreFacade) {
    this.__storage = 'store' in clientOrStore ? new SparqlStoreFacade(clientOrStore) : clientOrStore
    this.__resources = new TermMap()
    this.__deletedGraphs = new TermSet()
  }

  async get(id: string | Term | undefined, opts?: GetOptions): Promise<GraphPointer<NamedNode, DatasetExt>> {
    let resource: GraphPointer<NamedNode, DatasetExt> | undefined
    let term: NamedNode | undefined
    if (typeof id === 'string') {
      term = $rdf.namedNode(id)
    } else if (id && id.termType !== 'NamedNode') {
      throw new Error('Loaded resource id must be NamedNode')
    } else {
      term = id
    }

    if (term) {
      if (!this.__resources.has(term)) {
        resource = await this.__storage.loadResource(term)
        if (resource) {
          this.__resources.set(term, resource)
        }
      }

      resource = this.__resources.get(term)
    }

    if (!resource) {
      if (opts?.allowMissing) {
        return undefined as any
      }
      throw new NotFoundError(term)
    }

    return resource
  }

  async getResource<T extends RdfResourceCore>(id: string | Term | undefined, opts?: GetOptions):Promise<T | undefined> {
    const pointer = await this.get(id, opts)
    if (!pointer) return undefined

    return RdfResource.factory.createEntity<T>(pointer)
  }

  async save(): Promise<void> {
    await this.__storage.writeChanges(this.__resources, this.__deletedGraphs)
    this.__deletedGraphs.clear()
  }

  create(id: NamedNode, { implicitlyDereferencable = true }: ResourceCreationOptions = {}): GraphPointer<NamedNode, DatasetExt> {
    if (id.termType !== 'NamedNode') {
      throw new Error('Resource must be identified by a NamedNode')
    }

    if (this.__resources.has(id)) {
      throw new Error(`Resource <${id.value}> already exists`)
    }

    const pointer = cf({ dataset: $rdf.dataset(), term: id })

    if (implicitlyDereferencable) {
      pointer.addOut(rdf.type, hydra.Resource)
    }

    this.__resources.set(id, pointer)
    return pointer
  }

  async createMember(collectionId: NamedNode, id: NamedNode, options?: ResourceCreationOptions): Promise<GraphPointer<NamedNode, DatasetExt>> {
    const member = this.create(id, options)
    const collection = await this.get(collectionId)

    if (!collection) {
      throw new Error(`Collection <${collectionId}> not found`)
    }

    collection.out(hydra.manages).forEach((manages) => {
      const property = manages.out(hydra.property).term
      const subject = manages.out(hydra.subject).term
      const object = manages.out(hydra.object).term

      if (object && subject) {
        warn('hydra:manages cannot have both subject and object. Got %s and %s respectively', subject.value, object.value)
        return
      }

      if (property && object) {
        member.addOut(property, object)
      } else if (property && subject) {
        member.addIn(property, subject)
      }
    })

    return member
  }

  delete(id: NamedNode): void {
    this.__deletedGraphs.add(id)
    this.__resources.delete(id)
  }
}
