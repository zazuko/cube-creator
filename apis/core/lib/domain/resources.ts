import ResourceStoreImpl, { ResourceStore } from '../ResourceStore'
import { streamClient } from '../query-client'

export function resourceStore(): ResourceStore {
  return new ResourceStoreImpl(streamClient)
}
