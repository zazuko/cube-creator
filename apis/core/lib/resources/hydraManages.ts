import { NamedNode } from 'rdf-js'
import { GraphPointer } from 'clownface'
import { hydra } from '@tpluscode/rdf-ns-builders'

type ManagesBlock = {
  object: NamedNode
  property: NamedNode
} | {
  subject: NamedNode
  property: NamedNode
}

export function manages(manages: ManagesBlock): [NamedNode, (manages: GraphPointer) => void] {
  return [hydra.manages, (managesPointer: GraphPointer) => {
    managesPointer.addOut(hydra.property, manages.property)

    if ('subject' in manages) {
      managesPointer.addOut(hydra.subject, manages.subject)
    } else if ('object' in manages) {
      managesPointer.addOut(hydra.object, manages.object)
    }
  }]
}
