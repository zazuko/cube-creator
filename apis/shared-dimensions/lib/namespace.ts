import namespace from '@rdfjs/namespace'
import env from './env'

type Shapes = 'shape/shared-dimension-create'
| 'shape/shared-dimension-update'
| 'shape/shared-dimension-term-create'
| 'shape/shared-dimension-term-update'
| 'shape/hierarchy-create'
| 'shape/hierarchy'

export const shape = namespace<Shapes>(`${env.MANAGED_DIMENSIONS_BASE}dimension/_`)
