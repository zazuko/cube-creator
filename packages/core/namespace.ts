import prefixes from '@zazuko/rdf-vocabularies/prefixes'
import namespace from '@rdf-esm/namespace'

export { shape } from './namespaces/shapes'

type CubeCreatorClass =
  'CubeProject' |
  'ProjectsCollection' |
  'CsvMapping' |
  'EntryPoint' |
  'CSVSource' |
  'CSVSourceCollection' |
  'Table' |
  'TableCollection' |
  'ObservationTable' |
  'ColumnMapping' |
  'LiteralColumnMapping' |
  'ReferenceColumnMapping' |
  'CreateLiteralColumnMappingAction' |
  'CreateReferenceColumnMappingAction' |
  'Job' |
  'JobCollection' |
  'TransformAction' |
  'ImportAction' |
  'PublishAction' |
  'UnlistAction' |
  'DimensionMetadataCollection' |
  'SharedDimensionsCollection' |
  'Observations' |
  'TransformJob' |
  'ImportJob' |
  'PublishJob' |
  'UnlistJob' |
  'ReplaceCSVAction' |
  'MediaLocal' |
  'MediaURL' |
  'OriginalValuePredicate'

type CubeCreatorProperty =
  'projectSourceKind' |
  'projects' |
  'csvMapping' |
  'csvSource' |
  'csvSourceCollection' |
  'csvColumnSample' |
  'sourceKind' |
  'csvw' |
  'tables' |
  'identifierTemplate' |
  'project' |
  'dataset' |
  'namespace' |
  'cubeGraph' |
  'columnMapping' |
  'sourceColumn' |
  'targetProperty' |
  'identifierMapping' |
  'referencedTable' |
  'referencedColumn' |
  'datatype' |
  'language' |
  'defaultValue' |
  'isObservationTable' |
  'cube' |
  'jobCollection' |
  'dimensionMetadata' |
  'observations' |
  'publishGraph' |
  'latestPublishedRevision' |
  'revision' |
  'dimensionMapping' |
  'sharedDimension' |
  'sharedDimensions' |
  'applyMappings' |
  'dimensionType' |
  'CubeProject/sourceCube' |
  'CubeProject/sourceEndpoint' |
  'CubeProject/sourceGraph' |
  'export' |
  'projectDetails' |
  'validationReport'

type OtherTerms =
  'dash' |
  'projectSourceKind/CSV' |
  'projectSourceKind/ExistingCube' |
  'projectSourceKind/ExportedProject'

type MetaDataProperty =
  'publishOnOpendata' |
  'publishOnVisualize'

type CubeCreatorTerms = CubeCreatorClass | CubeCreatorProperty | OtherTerms | MetaDataProperty

type SharedDimensionsTerms =
  'SharedDimension' |
  'SharedDimensions' |
  'SharedDimensionTerm' |
  'SharedDimensionTerms' |
  'SharedDimensionExport' |
  'sharedDimensions' |
  'sharedDimension' |
  'terms' |
  'onlyValidTerms' |
  'export' |
  'createAs' |
  'dynamicPropertyType' |
  'hierarchies' |
  'Hierarchies' |
  'Hierarchy'

type MetaTerms =
  'SharedDimension' |
  'dataKind' |
  'DimensionRelation' |
  'dimensionRelation' |
  'relatesTo' |
  'hasHierarchy' |
  'hierarchyRoot' |
  'nextInHierarchy' |
  'Hierarchy'

prefixes.cube = 'https://cube.link/'
prefixes.view = 'https://cube.link/view/'
prefixes.meta = 'https://cube.link/meta/'
prefixes.relation = 'https://cube.link/relation/'

export const query = namespace('http://hypermedia.app/query#')
export const cube = namespace(prefixes.cube)
export const meta = namespace<MetaTerms>(prefixes.meta)
export const view = namespace(prefixes.view)
export const relation = namespace(prefixes.relation)
export const hydraBox = namespace('http://hydra-box.org/schema/')
export const cc = namespace<CubeCreatorTerms>('https://cube-creator.zazuko.com/vocab#')
export const md = namespace<SharedDimensionsTerms>('https://cube-creator.zazuko.com/shared-dimensions/vocab#')
export const editor = namespace(cc.dash.value)
export const sh1 = namespace('https://forms.hypermedia.app/shaperone#')
export const hex = namespace('https://w3id.org/hydra/extension#')
export const iso6391 = namespace('http://lexvo.org/id/iso639-1/')
