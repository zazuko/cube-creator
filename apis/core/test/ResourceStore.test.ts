import { describe, it, beforeEach } from 'mocha'
import { expect } from 'chai'
import sinon from 'sinon'
import $rdf from 'rdf-ext'
import { StreamClient } from 'sparql-http-client/StreamClient'
import StreamQuery from 'sparql-http-client/StreamQuery'
import StreamStore from 'sparql-http-client/StreamStore'
import ResourceStore from '../lib/ResourceStore'
import { ex } from './support/namespace'
import { hydra, rdf, schema } from '@tpluscode/rdf-ns-builders'
import { manages } from '../lib/resources/hydraManages'

describe('ResourceStore', () => {
  let client: StreamClient
  let query: sinon.SinonStubbedInstance<StreamQuery>
  let store: sinon.SinonStubbedInstance<StreamStore>

  beforeEach(() => {
    query = sinon.createStubInstance(StreamQuery)
    store = sinon.createStubInstance(StreamStore)

    client = {
      query,
      store: store as any,
    }
  })

  describe('get', () => {
    it('loads resource from sparql endpoint', async () => {
      // given
      const store = new ResourceStore(client)
      query.construct.resolves($rdf.dataset().toStream())

      // when
      await store.get(ex.Foo)

      // then
      expect(client.query.construct).to.have.been.calledOnce
    })

    it('called twice returns same object', async () => {
      // given
      const store = new ResourceStore(client)
      query.construct.resolves($rdf.dataset().toStream())
      const expected = await store.get(ex.Foo)

      // when
      const actual = await store.get(ex.Foo)

      // then
      expect(actual).to.eq(expected)
    })

    it('returns object previously created', async () => {
      // given
      const store = new ResourceStore(client)
      const expected = store.create(ex.Foo)

      // when
      const actual = await store.get(ex.Foo)

      // then
      expect(actual).to.eq(expected)
      expect(client.query.construct).to.not.have.been.called
    })
  })

  describe('create', () => {
    it('throws when called twice', () => {
      // given
      const store = new ResourceStore(client)

      // when
      store.create(ex.Foo)

      // then
      expect(() => store.create(ex.Foo)).to.throw
    })

    it('adds hydra:Resource type by default', () => {
      // given
      const store = new ResourceStore(client)

      // when
      const foo = store.create(ex.Foo)

      // then
      expect(foo.out(rdf.type).term).to.deep.eq(hydra.Resource)
    })

    it('does not add hydra:Resource type when flag is false', () => {
      // given
      const store = new ResourceStore(client)

      // when
      const foo = store.create(ex.Foo, {
        implicitlyDereferencable: false,
      })

      // then
      expect(foo.out(rdf.type).terms).to.have.length(0)
    })
  })

  describe('createMember', () => {
    it('asserts triples using hydra:manages patterns', async () => {
      // given
      const store = new ResourceStore(client)
      store.create($rdf.namedNode('people'))
        .addOut(...manages({
          property: rdf.type,
          object: schema.Person,
        }))
        .addOut(...manages({
          property: schema.employee,
          subject: ex.Company,
        }))

      // when
      const person = await store.createMember($rdf.namedNode('people'), ex.Foo)

      // then
      expect(person.out(rdf.type).terms).to.deep.contain(schema.Person)
      expect(person.in(schema.employee).term).to.deep.equal(ex.Company)
    })

    it('does not assert anything when hydra:manages are malformed', async () => {
      // given
      const store = new ResourceStore(client)
      store.create($rdf.namedNode('people'))
        .addOut(hydra.manages, manages => {
          manages.addOut(hydra.property, rdf.type)
        })
        .addOut(hydra.manages, manages => {
          manages
            .addOut(hydra.subject, ex.Company)
            .addOut(hydra.property, schema.employee)
            .addOut(hydra.object, ex.Bogus)
        })

      // when
      const person = await store.createMember($rdf.namedNode('people'), ex.Foo, {
        implicitlyDereferencable: false,
      })

      // then
      expect(person.out().terms).to.have.length(0)
      expect(person.in().terms).to.have.length(0)
    })
  })

  describe('save', () => {
    it('stores resources in individual named graphs', async function () {
      // given
      const resources = new ResourceStore(client)
      resources.create(ex.Foo).addOut(ex.foo, 'foo')
      resources.create(ex.Bar).addOut(ex.bar, 'bar')

      // when
      await resources.save()

      // then
      expect(client.query.update).to.have.been.calledOnce
      expect(query.update.firstCall.args[0]).to.matchSnapshot(this)
    })
  })

  describe('delete', () => {
    it('delete is adding DROP GRAPH statement and removes from inserts', async function () {
      // given
      const resourceStore = new ResourceStore(client)
      resourceStore.create(ex.Foo)
      resourceStore.delete(ex.Foo)

      // when
      await resourceStore.save()

      // then
      expect(query.update.firstCall.args[0]).to.matchSnapshot(this)
    })
  })
})
