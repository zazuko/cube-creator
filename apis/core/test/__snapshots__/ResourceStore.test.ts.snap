exports["ResourceStore save stores resources in individual named graphs"] = "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\nPREFIX hydra: <http://www.w3.org/ns/hydra/core#>\n\nDROP GRAPH <http://example.com/Foo>;\nDROP GRAPH <http://example.com/Bar>;\n\nINSERT DATA {\n  \n  \nGRAPH <http://example.com/Foo> {\n        \n<http://example.com/Foo> rdf:type hydra:Resource .\n<http://example.com/Foo> <http://example.com/foo> \"foo\" .\n      }\n  \nGRAPH <http://example.com/Bar> {\n        \n<http://example.com/Bar> rdf:type hydra:Resource .\n<http://example.com/Bar> <http://example.com/bar> \"bar\" .\n      }\n}";

exports["ResourceStore delete delete is adding DROP GRAPH statement and removes from inserts"] = "DROP GRAPH <http://example.com/Foo>;\n\nINSERT DATA {\n  \n}";

